File: Hot 100/Subarray/0560_SubarraySumEqualsK.java
// Two Traversal:
class Solution {
    public int subarraySum(int[] nums, int k) {
        int[] sums = new int[nums.length + 1];
        for (int i = 0; i < nums.length; i++) {
            sums[i + 1] = sums[i] + nums[i];
        }
        Map<Integer, Integer> map = new HashMap<>();
        int ans = 0;
        for (int sum : sums) {
            ans += map.getOrDefault(sum - k, 0);
            map.merge(sum, 1, Integer::sum);
        }
        return ans;
    }
}

// One Traversal:
class Solution {
    public int subarraySum(int[] nums, int k) {
        int sum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        int ans = 0;
        map.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            ans += map.getOrDefault(sum - k, 0);
            map.merge(sum, 1, Integer::sum);
        }
        return ans;
    }
}

File: Top 100 Liked/Subarray/0560_subarray_sum_equals_k.py
# Two Traversal
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        occ: dict[int, int] = {}
        ans = 0
        n = len(nums)
        sums = [0] * (n + 1)
        for i in range(n):
            sums[i + 1] = nums[i] + sums[i]
        for i in range(n + 1):
            ans = ans + occ.get(sums[i] - k, 0)
            occ[sums[i]] = occ.get(sums[i], 0) + 1
        return ans

# One Traversal
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        ans = 0
        occ: dict[int, int] = {}
        n = len(nums)
        sums = 0
        occ[0] = 1
        for i in range(n):
            sums = sums + nums[i]
            ans = ans + occ.get(sums - k, 0)
            occ[sums] = occ.get(sums , 0) + 1
        return ans

# simpler solution:
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        ans = 0
        presum = 0
        dic = defaultdict(int)
        dic[0] = 1

        for i in nums:
            presum += i
            ans += dic[presum - k]
            dic[presum] += 1

        return ans
